OLM Integration

avisiedo opened the pull request:
As an administrator I want to integrate freeipa-operator with OLM that meet with:
- Split the service accounts in one for the controller and one more for the freeipa workload, shrinking more the permissions that each component has in the cluster.
- RBAC resources creation/deletion are delegated to OLM, letting the operator controller run with less permissions.
- Restore Makefile and extract the additional rules in external included Makefiles.
- Clean-up the usage of RBAC resource from the controller code.
- Create the OLM bundle.
- OLM will manage the operator lifecycle, letting the user install/uninstall the operator from the OpenShift console.

avisiedo committed:
Decouple samples from Makefile

avisiedo committed:
Decouple go-get-tool from Makefile

avisiedo committed:
Remove kind rules from Makefile

avisiedo committed:
Update vendor and go.mod for v3 project version

avisiedo committed:
Update vendor: add missing go.sum file

avisiedo committed:
Add testbin/ to gitignore

avisiedo committed:
Update project to v3

avisiedo committed:
Add health and ready checks; Update controller arguments; Update unit tests

avisiedo committed:
Update crd manifests

avisiedo committed:
Update rbac manifests and use different serviceaccounts for controller and workload

avisiedo committed:
Comment rbac object management and update serviceaccountname for the workload

avisiedo committed:
Update webhook manifests

avisiedo committed:
Update manager manifests

avisiedo committed:
Update patch for auth_proxy

avisiedo committed:
Update prometheus monitoring to be aligned to operator-sdk

avisiedo committed:
Add controller manager configuration resource

avisiedo committed:
Add bundle manifest configuration

avisiedo committed:
Add scorecards

avisiedo committed:
Add private.mk to gitignore

avisiedo committed:
Decouple Makefile into addons and checks

avisiedo committed:
Restore default Makefile and inject changes from external files

avisiedo committed:
Update samples.mk documentation and force sample-recreate to not execute in parallel

avisiedo committed:
Add bundle for olm

avisiedo committed:
Fix phony target

avisiedo committed:
Set default goal to 'build'

avisiedo committed:
Update unit tests to match the service accunt name that is used for the workload

avisiedo committed:
Ignore bundle_tmp* directories

avisiedo committed:
Include cert-manager rules and update documentation

avisiedo committed:
Update documentation

avisiedo committed:
Add check for IMAGE_TAG_BASE

avisiedo committed:
Add scorecard rules

avisiedo committed:
Add printvars rule

avisiedo committed:
Update addons makefile

avisiedo committed:
Fix error message when cert-manager is not installed

avisiedo committed:
Ignore Visual Studio code files

avisiedo committed:
Handle deprecated and obsoletes rules after sync with the operator-sdk Makefile

avisiedo committed:
Set default rule 'help'

avisiedo committed:
Add deprecated rules into addons.mk

avisiedo committed:
Move config/samples/ to samples/ directory

avisiedo committed:
Move addons to the end, isolate more the go-get-tool macro and add support for local operator-sdk

avisiedo committed:
Simplify cert-manager-* rules

avisiedo committed:
Remove container-push from help

avisiedo committed:
Isolate tool installation macro

avisiedo committed:
Improve sample-create rule to create secret only when it does not exist, change sample-delete order and remove sample-build and sample-recreate from help

avisiedo committed:
Add scorecard rules to launch the scorecard tests

avisiedo committed:
Add operator-sdk rule to download it locally

avisiedo committed:
Decouple cert-manager configuration from any namespace

avisiedo committed:
Remove samples/ directory from manifest resources

avisiedo committed:
Add dnses-role to the controller-manager service account

avisiedo committed:
Update resrource names to match with the generated ones for the workload

avisiedo committed:
Drop AIM permissions, and leave only the necessary to use the service account associated to the controller

avisiedo committed:
Update the serviceaccount name associated to the workload

avisiedo committed:
Add macro to post-process configuration manifests

avisiedo committed:
Udate controller-manager role and rolebind

avisiedo committed:
Update unit tests to match the service account name

avisiedo committed:
Fix rbac rule annotations, clean-up AIM resources and fix some log messages

avisiedo committed:
Add kubectl-wait-for-value macro

avisiedo committed:
Update default channel at bundle.Dockerfile

avisiedo committed:
Update rule to wait cert-manager is installed

avisiedo committed:
Update default channel in annotations

avisiedo committed:
Ignore 'bundle/' directory which is generated by 'make bundle'

avisiedo committed:
Remove ignore for bundlet/ directory

avisiedo committed:
Ignore bundle-* directories

avisiedo committed:
Add current bundle state

avisiedo committed:
Add bundle-install and bundle-uninstall rules for automating the OLM lifecycle

avisiedo committed:
Add support to indicate a specific test

avisiedo committed:
Remove no needed output when running sample-delete

avisiedo committed:
Update clusterversionservice resource

avisiedo committed:
Add icon to the bundle

avisiedo committed:
Simplify rule for uninstalling the operator by using OLM

avisiedo committed:
Clean-up rules

avisiedo committed:
Clean-up makefile used to download specific operator-sdk

avisiedo committed:
Fix PROJECT_DIR when multiple makefiles are included

avisiedo committed:
Fix the rule fails when bin/ directory does not exist

avisiedo committed:
Add rules for scorecard-build and scorecard-push for the scorecard images

avisiedo committed:
Add .PHONY for scorecard-bundle and scorecard-image rules

avisiedo committed:
Add patches for generating bundle with the decriptors, add samples for alm-examples annotation and decouple the icon information in one patch

avisiedo committed:
Update IDMSpec comments, cleanup some scafoding comments, and update CRD from it

avisiedo committed:
Delete bundle/ directory which is completely generated by 'make bundle'

avisiedo committed:
Remove unnecessary association to the service account, and update the security context constraint resource name

avisiedo committed:
Comment SecurityContextConstraint resource

avisiedo committed:
Update namespace to the current one when calling 'make bundle'

avisiedo committed:
Update documentation to deploy with OLM from cli

avisiedo committed:
Ignore bundle/ directory which is completely generated by 'make bundle' rule

avisiedo committed:
Update README documentation

avisiedo committed:
Add yml2json and json2yaml tool installation

avisiedo committed:
linter: Fix hints for the README markdown file

avisiedo committed:
Remove yml2json

avisiedo committed:
Fix error message when yaml2json is not found

avisiedo committed:
pipeline:Fix pipeline jobs

- Add yaml2json dependency
- Make available IMAGE_TAG_BASE env var
- Update build job

avisiedo committed:
doc: Update documentation for the pipeline

avisiedo committed:
Remove wrong await invokation

avisiedo committed:
makefile: Add rules tidy, vendor and lint

avisiedo committed:
lint:Fix linter hints and add some generated files to ignore list

avisiedo committed:
makefile:Restore go-get-tool macro

avisiedo committed:
Add .venv rule to download python dependencies

avisiedo committed:
Use .venv for the pyton dependencies

avisiedo commented:
/retest

avisiedo committed:
Update description for host field in the custom resource definition

avisiedo committed:
Remove checks as the tools are installed in the macro if they are needed

avisiedo committed:
Update printvars rule to display the variable value without expand it

avisiedo committed:
Update WATCH_NAMESPACE to be defined only when KUBECONFIG is not empty

avisiedo committed:
pipeline: Fix github actions

- Add env file for main repo ci build.
- Update github-actions to force to create
  the ENV secret when no working in the main
  pipeline.
- Add ENV secret information and short
  description for container images
- Install python dependencies.
- Disable 'Install dependencies' step.
- Add rule to download dive tool from the Makefile.
- Rename .dive-ci.yaml to the default name
  that is used by dive when CI=true and update
  command to remove no needed option.
- Ignore generated dive files.
- Increase integration with make rules.
- Remove implicit 'test' rule when running
  the 'doker-build' rule.
- Disable integration tests when 'GITHUB_ACTIONS=true'.
- Cache bin/ tools.
- Cache '.venv' directory
- Remove 'Install dependencies' job

avisiedo committed:
linguist:Add 'vendor/' and rbac generated files

avisiedo committed:
pipeline:Fix ci-operator side and some improvements

- Add '-mod vendor' to build, run and test commands
- Ignore .venv directory
- Tidy dependencies
- Set GOCACHE to /tmp/.cache when runing in OPENSHIFT_CI
- Add links to the pipeline documentation.
- Set XDG_CACHE_HOME to /tmp/.cache when OPENSHIFT_CI=true
- Set USE_EXISTING_CLUSTER=1 when OPENSHIFT_CI=true
- Set HOME to /tmp when OPENSHIFT_CI=true
- Disable tests just for the controller package in both pipelines
- Add verbose mode to the test output

avisiedo commented:
/retest

avisiedo commented on .gitattributes:
```diff
@@ -0,0 +1,3 @@
+vendor/**                      linguist-vendored
```
This avoid that GitHub load the diffs for files on this directory.

tiran replied:
``linguist-vendored`` does not exclude files from diffs. GH only excludes the files from stats.You have to use ``linguist-generated`` to hide the diffs.

avisiedo commented on .gitattributes:
```diff
@@ -0,0 +1,3 @@
+vendor/**                      linguist-vendored
+config/rbac/role.yaml          linguist-generated
+config/rbac/role_binding.yaml  linguist-generated
```
After fix the annotations in the code, this files are generated by `controller-gen`, so they are marked as generated in the repository.

avisiedo commented on .github/workflows/main.yml:
```diff
@@ -23,18 +23,60 @@ jobs:
+
+      - name: Prepare environment variables
+        run: |
+          if [ "${{ github.repository }}" != "freeipa/freeipa-operator" ]; then
+            if [ "${{ secrets.ENV }}" == "" ]; then
+              echo "ERROR:You need to define ENV secret with your environment."
+              echo "INFO:See DEVOPS.md file for further information"
+              cat DEVOPS.md
+              exit 1
+            else
+              [ -e "bin" ] || mkdir bin
+              cat > env.ci <<EOF
+          ${{ secrets.ENV }}
+          EOF
+            fi
+          else
+            cp -vf build/ci/env env.ci
+          fi
```
Prepare the `env.ci` file; if it is the main repository, it is copied from 'build/ci/env' reopsitory file. For a forked repository, the content of the file with the customs are loaded from the `ENV` secret.

avisiedo commented on .github/workflows/main.yml:
```diff
@@ -23,18 +23,60 @@ jobs:
+          [ -e "~/go/pkg/mod" ] || mkdir -p ~/go/pkg/mod
+
+      # yamllint disable-line rule:line-length
+      - run: env -i bash -c 'set -a; source env.ci; env --unset=PWD --unset=SHLVL --unset=_' >> "$GITHUB_ENV"
```
Load environment variables from `env.ci` file, making them available for the next steps in the job. This is repeated for each job.

avisiedo commented on .github/workflows/main.yml:
```diff
@@ -23,18 +23,60 @@ jobs:
+
+      # yamllint disable-line rule:line-length
+      - run: env -i bash -c 'set -a; source env.ci; env --unset=PWD --unset=SHLVL --unset=_' >> "$GITHUB_ENV"
+      - uses: actions/upload-artifact@v3
+        with:
+          name: env
+          path: env.ci
```
Upload the `env.ci` as an artifact. This is used to share it into the different jobs, so it can be just downloaded and used to define the needed environment variables.

avisiedo committed:
Comment the web reference

frasertweedale commented on Makefile:
```diff
@@ -360,3 +170,62 @@ GOBIN=$(PROJECT_DIR)/bin go get $(2) ;\
 }
 endef
+
+.PHONY: bundle
+bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metadata, then validate generated files.
+	$(OPERATOR_SDK) generate kustomize manifests -q
+	cd config/default && $(KUSTOMIZE) edit set namespace $(WATCH_NAMESPACE)
+	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	$(KUSTOMIZE) build config/manifests | $(OPERATOR_SDK) generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)
+	$(OPERATOR_SDK) bundle validate ./bundle
+
+.PHONY: bundle-build
+bundle-build: ## Build the bundle image.
+	docker build -f bundle.Dockerfile -t $(BUNDLE_IMG) .
+
+.PHONY: bundle-push
+bundle-push: ## Push the bundle image.
+	$(MAKE) docker-push IMG=$(BUNDLE_IMG)
```
Review comment: it probably makes sense to have dependencies between these targets; bundle-push <- bundle-build <- bundle.

avisiedo replied:
Updated!

frasertweedale commented on Makefile:
```diff
@@ -360,3 +170,62 @@ GOBIN=$(PROJECT_DIR)/bin go get $(2) ;\
+# Build a catalog image by adding bundle images to an empty catalog using the operator package manager tool, 'opm'.
+# This recipe invokes 'opm' in 'semver' bundle add mode. For more information on add modes, see:
+# https://github.com/operator-framework/community-operators/blob/7f1438c/docs/packaging-operator.md#updating-your-existing-operator
+.PHONY: catalog-build
+catalog-build: opm ## Build a catalog image.
+	$(OPM) index add --container-tool docker --mode semver --tag $(CATALOG_IMG) --bundles $(BUNDLE_IMGS) $(FROM_INDEX_OPT)
+
+# Push the catalog image.
+.PHONY: catalog-push
+catalog-push: ## Push a catalog image.
+	$(MAKE) docker-push IMG=$(CATALOG_IMG)
```
Again, probably makes sense to have dependency here too?

avisiedo replied:
Updated!

frasertweedale commented on README.md:
```diff
@@ -24,46 +24,50 @@ Experimental freeipa-operator for Freeipa.
+   export IMAGE_TAG_BASE=quay.io/USER_ORG/freeipa-operator
    podman login quay.io
-   make container-push IMG=quay.io/USER_ORG/freeipa-operator:dev-test
+   make container-build
+   make container-push
```
update to docker-build / docker-push

frasertweedale commented on README.md:
```diff
@@ -74,10 +78,58 @@ Experimental freeipa-operator for Freeipa.
+
+----
+
 <!-- TODO When the read of ingresDomain is implemented, remove the
           block below. -->
```
Remove obsolete comment.

frasertweedale commented on bundle.Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM scratch
```
One thing: although this file is autogenerated, it would be good to have a documentation (or a link) that explains what all the labels means.

avisiedo replied:
I have added a link in readme.md file when the `make bundle` step, to the documentation about the labels.

avisiedo replied:
- https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md#bundle-annotations
- https://olm.operatorframework.io/docs/tasks/creating-operator-bundle/#contents-of-annotationsyaml-and-the-dockerfile

frasertweedale reviewed (commented):
Review 25% complete. Publishing the comments so far.

avisiedo committed:
Update 'docker-*' rules in documentation

avisiedo committed:
Remove obsolete comment

avisiedo committed:
Add more information and note about remaining PVC resource

avisiedo committed:
Restore whitespace

avisiedo committed:
Restore whitespaces in comments

avisiedo committed:
Restore whitespaces in comments

avisiedo committed:
Add reference to documentation about the labels and annotations from the operator bundle

avisiedo committed:
Fix default controller image to point out to freeipa organization

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+      name: idms.idmocp.redhat.com
+      version: v1alpha1
+  description: Deploy Freeipa in OpenShift
+  displayName: Identity Management as a Service
```
Name to display in OLM installed operators and in OLM catalog from the OpenShift console.

avisiedo replied:
![image](https://user-images.githubusercontent.com/3994774/161236993-51fa9ffc-b6ac-4a80-8483-e39ccab62cb7.png)

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+      kind: IDM
+      name: idms.idmocp.redhat.com
+      version: v1alpha1
+  description: Deploy Freeipa in OpenShift
```
Description to display in OLM installed operators and in OLM catalog from the OpenShift console.

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+  displayName: Identity Management as a Service
+  install:
+    spec:
+      clusterPermissions:
```
Generated from the `config/rbac` directory

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+        serviceAccountName: idm-operator-controller-manager
+      deployments: []
+      permissions:
+      - rules:
```
Generated from the `config/rbac` directory

avisiedo commented on config/manifests/kustomization.yaml:
```diff
@@ -0,0 +1,59 @@
+  - bases/freeipa-operator.clusterserviceversion.yaml
+  - ../default
+  - ../scorecard
+  - ../samples
```
This is translated to the `alm-examples` annotation when the `bundle` directory is generated by `make bundle`.

avisiedo commented on config/manifests/kustomization.yaml:
```diff
@@ -0,0 +1,59 @@
+  - ../samples
+
+patchesJson6902:
+  - target:
+      group: operators.coreos.com
+      version: v1alpha1
+      kind: ClusterServiceVersion
+      name: freeipa-operator.v0.0.0
+      namespace: placeholder
+    path: patches/icon.yaml
```
Inject the icon to be displayed in OLM OpenShift console for our operator.

avisiedo committed:
Restore the order of the comments, which was lost when 'make bundle' was invoked

avisiedo committed:
Update repository url

avisiedo committed:
Set vendor linguist to generated

avisiedo committed:
Add link to design doc about the bundle annotations

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+    type: MultiNamespace
+  - supported: false
+    type: AllNamespaces
+  keywords:
+  - freeipa
+  - kerberos
+  - pki
+  - sso
+  - idm
```
Double check list of keywords: what to delete, what to add, what to update?

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+  - pki
+  - sso
+  - idm
+  links:
+  - name: Freeipa Operator
+    url: https://github.com/freeipa/freeipa-operator
```
Double check if this is the URL to display here. It is displayed into the operator information in OLM from the OpenShift console.

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+  - name: Freeipa Operator
+    url: https://github.com/freeipa/freeipa-operator
+  maintainers:
+  - email: freeipa-users@lists.fedorahosted.org
+    name: freeipa-users
```
Double check if this is the email to indicate here.

avisiedo commented on config/manifests/bases/freeipa-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,157 @@
+  provider:
+    name: Red Hat
+    url: https://www.redhat.com
+  version: 0.0.0
```
![image](https://user-images.githubusercontent.com/3994774/161236123-56196b36-51e1-4c7c-98ae-06df49464436.png)

avisiedo commented on devel/lint.ignore:
```diff
@@ -2,3 +2,23 @@ miscelanea/003-scc-and-caps/README.md
 miscelanea/004-route-traffic/README.md
 config/manager/kustomization.yaml
 config/webhook/manifests.yaml
+config/manifests/bases/freeipa-operator.clusterserviceversion.yaml
+config/rbac/role.yaml
+config/rbac/role_binding.yaml
+config/default/kustomization.yaml
+bundle/tests/scorecard/config.yaml
+bundle/manifests/freeipa-operator.clusterserviceversion.yaml
+bundle/manifests/idm-operator-controller-manager-metrics-service_v1_service.yaml
+bundle/manifests/idm-operator-freeipa-role_rbac.authorization.k8s.io_v1_role.yaml
+bundle/manifests/idm-operator-freeipa-rolebinding_rbac.authorization.k8s.io_v1_rolebinding.yaml
+bundle/manifests/idm-operator-freeipa_v1_serviceaccount.yaml
+bundle/manifests/idm-operator-leader-election-role_rbac.authorization.k8s.io_v1_role.yaml
+bundle/manifests/idm-operator-leader-election-rolebinding_rbac.authorization.k8s.io_v1_rolebinding.yaml
+bundle/manifests/idm-operator-manager-config_v1_configmap.yaml
+bundle/manifests/idm-operator-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml
+bundle/manifests/idm-operator-proxy-role_rbac.authorization.k8s.io_v1_clusterrole.yaml
+bundle/manifests/idm-operator-proxy-rolebinding_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml
+bundle/manifests/idm-operator-webhook-service_v1_service.yaml
+bundle/manifests/idmocp.redhat.com_idms.yaml
+bundle/metadata/annotations.yaml
+bundle/tests/tests/scorecard/config.yaml
```
Add more generated files to the linting exclusion list

avisiedo commented on mk/cert-manager.mk:
```diff
@@ -1,22 +1,12 @@
 
 .PHONY: cert-manager-install
-cert-manager-install:
+cert-manager-install:  ## Install cert-manager operator
 	kubectl create -f config/certmanager/subscription.yaml
+	kubectl wait Subscription/cert-manager -n openshift-operators --for=condition=CatalogSourcesUnhealthy=False
```
We need cert-manager operator when we deploy the controller in our development OpenShift cluster without OLM by doing 'make deploy'.

avisiedo commented on mk/macros.mk:
```diff
@@ -0,0 +1,20 @@
+ifndef post-manifests
+define post-manifests
+@[ -e .venv ] || $(MAKE) .venv
+@(source .venv/bin/activate; yaml2json config/rbac/role.yaml) | jq '.kind="Role"' > config/rbac/role.json \
+&& (source .venv/bin/activate; json2yaml config/rbac/role.json > config/rbac/role.yaml) \
+&& rm -f config/rbac/role.json \
+&& (source .venv/bin/activate; yaml2json config/rbac/role_binding.yaml) | jq '.kind="RoleBinding" | .roleRef.kind="Role"' > config/rbac/role_binding.json \
+&& (source .venv/bin/activate; json2yaml config/rbac/role_binding.json) > config/rbac/role_binding.yaml \
+&& rm -f config/rbac/role_binding.json
```
This is called at 'make manifests' to impact the less as possible into the Makefile that generate `operator-sdk`. 

This update the ClusterRole and ClusterRoleBindings for the controller manager to Role and RoleBindings resources to restrict the controller permissions to a namespaced scope instead of cluster wide. When the manifests are generated.

avisiedo committed:
build:Add bundle and catalog rule dependencies

avisiedo committed:
deploy:Add dependencies to deploy an updated operator with OLM

avisiedo committed:
Run 'bundle' rule with no parallel jobs

avisiedo committed:
repo:The CRD definition is generated by controller-gen; adding linguist-generated for that file

avisiedo commented on config/manifests/patches/spec-descriptors.yaml:
```diff
@@ -0,0 +1,18 @@
+- op: add
+  path: /spec/customresourcedefinitions/owned/0/specDescriptors
+  value:
+  - displayName: "Realm"
```
Add here the Host field.

avisiedo committed:
bundle:Add the host field to spec-descriptors

avisiedo committed:
bundle:Add docker-build dependency

docker-build is reused for other rules, so instead of
add a direct dependency at docker-push, it is add
just here in the 'make bundle-install' rule.
